MsCorePkg/MacAddressEmulationDxe/MacAddressEmulationDxe.c

```diff
@@ -0,0 +1,278 @@
+    // artificially lower to TPL_CALLBACK for this call to work as expected
+    OldTpl = gBS->RaiseTPL (TPL_HIGH_LEVEL);
+    gBS->RestoreTPL (TPL_CALLBACK);
+
+    Status = Snp->StationAddress (Snp, FALSE, &Context->EmulationAddress);
+
+    gBS->RaiseTPL (TPL_HIGH_LEVEL);
+    gBS->RestoreTPL (OldTpl);
+  }
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "[%a]: Failed to set MAC address on SNP interface. Status=%r\n", __FUNCTION__, Status));
+  } else {
+    // Update context to indicate that we've assigned the emulation to this particular device.
+    // save the permanent address to facilitate the check above.
+    CopyMem (&Context->PermanentAddress, &Snp->Mode->PermanentAddress, NET_ETHER_ADDR_LEN);
+    Context->Assigned = TRUE;
+  }
+
+  DEBUG ((DEBUG_VERBOSE, "[%a]: End\n", __FUNCTION__));
+}
+
+/**
+  Callback that is invoked when an SNP instance is Initialized. Checks the newly installed SNP registrations (if any)
+  and updates the MAC address if a supported adapter is found.
+
+  @param[in]  Event   - The EFI_EVENT for this notify.
+  @param[in]  Context - An instance of MAC_EMULATION_SNP_NOTIFY_CONTEXT.
+
+  @remark  This callback should be at least 1 level higher TPL than network stack to ensure timely mac programming
+           or network stack start should be delayed until later.
+**/
+VOID
+EFIAPI
+SimpleNetworkProtocolNotify (
+  IN EFI_EVENT  Event,
+  IN VOID       *Context
+  )
+{
+  MAC_EMULATION_SNP_NOTIFY_CONTEXT  *MacContext;
+  EFI_SIMPLE_NETWORK_PROTOCOL       *SnpToConfigureEmu;
+
+  DEBUG ((DEBUG_VERBOSE, "[%a]: Start\n", __FUNCTION__));
+
+  if (Context == NULL) {
+    DEBUG ((DEBUG_ERROR, "[%a]: Context unexpectedly null.\n", __FUNCTION__));
+    ASSERT (Context != NULL);
+    return;
+  }
+
+  MacContext = (MAC_EMULATION_SNP_NOTIFY_CONTEXT *)Context;
+
+  SnpToConfigureEmu = FindMatchingSnp (MacContext);
+
+  SetSnpMacViaContext (SnpToConfigureEmu, MacContext);
+
+  DEBUG ((DEBUG_VERBOSE, "[%a]: End\n", __FUNCTION__));
+}
+
+/**
+  Driver entry:  Initializes MAC Address Emulation
+
+  @param[in]  ImageHandle - Handle to this image.
+  @param[in]  SystemTable - Pointer to the system table.
+
+  @retval  EFI_STATUS code.
+**/
+EFI_STATUS
+EFIAPI
+MacAddressEmulationEntry (
+  IN EFI_HANDLE        ImageHandle,
+  IN EFI_SYSTEM_TABLE  *SystemTable
+  )
+{
+  MAC_EMULATION_SNP_NOTIFY_CONTEXT  *Context;
+  EFI_MAC_ADDRESS                   Address;
+  EFI_STATUS                        Status;
+
+  // Determine general platform runtime support.  Return unsupported to fully unload driver if not enabled.
+  Status = IsMacEmulationEnabled (&Address);
+  if (EFI_ERROR (Status)) {
+    if (Status != EFI_UNSUPPORTED) {
+      DEBUG ((DEBUG_ERROR, "[%a]: Failed to determine MAC Emulated Address support. Status = %r\n", __FUNCTION__, Status));
+    }
+
+    return Status;
+  }
+
+  // Allocate and initialize context
+  Context = AllocateZeroPool (sizeof (MAC_EMULATION_SNP_NOTIFY_CONTEXT));

```

Comment:
Consider using the variable name for the size. If somone updates the type in the future, they won't have to update the type here as well.

```suggestion
  Context = AllocateZeroPool (sizeof (*Context));
```